source/MessageLogPanel.cpp

```diff
@@ -0,0 +1,147 @@
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "MessageLogPanel.h"
+
+#include "text/alignment.hpp"
+#include "Color.h"
+#include "Command.h"
+#include "FillShader.h"
+#include "text/Font.h"
+#include "text/FontSet.h"
+#include "GameData.h"
+#include "Preferences.h"
+#include "Screen.h"
+#include "Sprite.h"
+#include "SpriteSet.h"
+#include "SpriteShader.h"
+#include "UI.h"
+#include "text/WrappedText.h"
+
+using namespace std;
+
+namespace {
+	const double PAD = 10.;
+	const double LINE_HEIGHT = 25.;
+}
+
+
+
+MessageLogPanel::MessageLogPanel(const vector<pair<string, Messages::Importance>> &messages)
+	: messages(messages)
+{
+	SetInterruptible(false);
+}
+
+
+
+// Draw this panel.
+void MessageLogPanel::Draw()
+{
+	// Dim out everything outside this panel.
+	DrawBackdrop();
+
+	// Draw the panel.
+	width = Screen::Width() - 300;
+	const Color &backColor = *GameData::Colors().Get("message log background");
+	FillShader::Fill(
+		Point(Screen::Left() + .5 * width, 0.),
+		Point(width, Screen::Height()),
+		backColor);
+
+	const Sprite *edgeSprite = SpriteSet::Get("ui/right edge");
+	if(edgeSprite->Height())
+	{
+		// If the screen is high enough, the edge sprite should repeat.
+		double spriteHeight = edgeSprite->Height();
+		Point pos(
+			Screen::Left() + width + .5 * edgeSprite->Width(),
+			Screen::Top() + .5 * spriteHeight);
+		for( ; pos.Y() - .5 * spriteHeight < Screen::Bottom(); pos.Y() += spriteHeight)
+			SpriteShader::Draw(edgeSprite, pos);
+	}
+
+	const Font &font = FontSet::Get(14);
+
+	// Parameters for drawing messages:
+	WrappedText messageLine(font);
+	messageLine.SetAlignment(Alignment::LEFT);
+	messageLine.SetWrapWidth(width - 2. * PAD);
+
+	// Draw messages.
+	Point pos = Screen::BottomLeft() + Point(PAD, scroll);
+	for(const auto &it : messages)
+	{
+		messageLine.Wrap(it.first);
+		pos.Y() -= messageLine.Height();
+		messageLine.Draw(pos, *Messages::GetColor(it.second, true));
+	}
+
+	maxScroll = max(0., scroll - pos.Y() + Screen::Top());
+}
+
+
+
+bool MessageLogPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &command, bool isNewPress)
+{
+	if(command.Has(Command::MESSAGE_LOG) || key == 'd' || key == SDLK_ESCAPE
+			|| (key == 'w' && (mod & (KMOD_CTRL | KMOD_GUI))))
+		GetUI()->Pop(this);
+	else if(key == SDLK_PAGEUP || key == SDLK_PAGEDOWN)
+	{
+		double direction = (key == SDLK_PAGEUP) - (key == SDLK_PAGEDOWN);
+		Drag(0., (Screen::Height() - 100.) * direction);
+	}
+	else if(key == SDLK_HOME || key == SDLK_END)
+	{
+		double direction = (key == SDLK_HOME) - (key == SDLK_END);
+		Drag(0., maxScroll * direction);
+	}
+	else if(key == SDLK_UP || key == SDLK_DOWN)
+	{
+		double direction = (key == SDLK_UP) - (key == SDLK_DOWN);
+		Drag(0., LINE_HEIGHT * direction);
+	}
+
+	return true;
+}
+
+
+
+bool MessageLogPanel::Click(int x, int y, int clicks)
+{
+	x -= Screen::Left();
+
+	if(x > width)
+		GetUI()->Pop(this);
+
+	return true;
+}
```

Comment:
not sure if we should do this, I dont know any other panel in game that closes if you click outside of it

Reply:
the Logbook, on which the Message Log is based

Reply:
Ah ok 👍